snuba/query/parser/expressions.py

```diff
@@ -61,44 +62,70 @@ def visit_function_name(self, node: Node, visited_children: Iterable[Any]) -> st
     def visit_column_name(self, node: Node, visited_children: Iterable[Any]) -> Column:
         return Column(None, None, node.text)
 
-    def visit_empty(
-        self, node: Node, visited_children: Iterable[Any]
-    ) -> Optional[Expression]:
-        return None
+    def visit_low_pri_op(self, node: Node, visited_children: Iterable[Any]) -> Any:
+        return node.text
+
+    def visit_high_pri_op(self, node: Node, visited_children: Iterable[Any]) -> Any:
+        return node.text
+
+    def visit_arithmetic_term(
+        self, node: Node, visited_children: Tuple[Any, Expression, Any]
+    ) -> Expression:
+        _, term, _ = visited_children
+
+        return term
 
     def visit_low_pri_arithmetic(
         self,
         node: Node,
-        visited_children: Tuple[Any, Expression, Any, Tuple[str, Expression]],
+        visited_children: Tuple[
+            Any, Expression, Any, Union[Any, Tuple[str, Expression]]
```

Comment:
```
Union[Any, Tuple[str, Expression]]
```
is basically the equivalent of 
```
Any
```
Since Any is a supertype.
Which is the case where you would receive a type that is not the Tuple? And what would that type be ?

Also when you iterate over `exp` that seems it should be typed as a Sequence, not a Tuple, since it may have any number of elements.

Reply:
Basically, if I understand correctly, there are 3 cases:

1) when the righthand child is "" (this gets bubbled up as a Node, since generic_visit)
2) when the righthand child is  [ [operator, Expression], [operator, Expression], ... ]
3) when the righthand child is  [operator, Expression]

So I would need some typing that can encapsulate these three situations. Currently, I have something like Union[Node, Tuple[Operator, Expression]]. I have tried modifying to include Sequence[Tuple[...]], List[Tuple[...]], but these cases are rejected by the type checker and are noted to violate Case 3.

Reply:
(And, just to confirm, lefthand child seems to be of type Expression, so this case should be fine)